### PR TITLE
Debian packaging via dh_virtualenv

### PR DESCRIPTION
**Pull request by hawkowl**

### CODE REVIEW

**Comment by richvdh** [comment]
Context for interested readers: this PR is based on a branch which copies the existing packaging from https://github.com/matrix-org/package-synapse-debian/tree/debian/debian. It supercedes #4212 and https://github.com/matrix-org/package-synapse-debian/pull/39.

The idea for now is that, because the changes here are a substantial rewrite, we will keep the existing python 2 packaging (in https://github.com/matrix-org/package-synapse-debian), and offer the packages built from this source as *alternatives* for those wanting to try out python 3.

File: debian/build_debian.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+set -e
+
+dpkg-buildpackage -us -uc -b
+
+cp ../*.deb /synapse/build/debs
```
Comment (richvdh) [inline debian/build_debian.sh]: this has lost its trailing NL

File: debian/build_debian.sh [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
```
Comment (richvdh) [inline debian/build_debian.sh]: could this have a comment saying what it does? It looks like it assumes it's running in a docker environment - could it say so?

File: debian/changelog [hunk 3]
```diff
@@ -1,3 +1,9 @@
+matrix-synapse-py3 (0.34.0) stretch; urgency=medium
+
+  * New upstream version 0.34.0
```
Comment (richvdh) [inline debian/changelog]: Since we're completely changing the debian packaging, it's probably worth mentioning that here, even though the info is to some extent duplicated in the main changelog.

File: debian/changelog [hunk 4]
```diff
@@ -627,4 +633,4 @@ matrix-synapse (0.9.2-1) wheezy; urgency=low
 
   * source package automatically created by stdeb 0.8.2
 
- -- Paul "LeoNerd" Evans <paul@matrix.org>  Fri, 12 Jun 2015 14:32:03 +0100
+ -- Paul "LeoNerd" Evans <paul@matrix.org>  Fri, 12 Jun 2015 14:32:03 +0100
```
Comment (richvdh) [inline debian/changelog]: please can you reinstate the lost NL

File: debian/control [hunk 5]
```diff
@@ -1,77 +1,39 @@
...
-Architecture: all
+Package: matrix-synapse-py3
+Architecture: amd64
+Conflicts: matrix-synapse
+Pre-Depends: dpkg (>= 1.16.1)
 Depends:
- ${misc:Depends},
- ${python:Depends},
  adduser,
- debconf,
```
Comment (richvdh) [inline debian/control]: I'm guessing we still depend on debconf?

File: debian/control [hunk 6]
```diff
@@ -1,77 +1,39 @@
...
- lsb-base (>= 3.0-6),
```
Comment (richvdh) [inline debian/control]: ... and probably this?

File: debian/control [hunk 7]
```diff
@@ -1,77 +1,39 @@
...
- convenience aliases like #matrix:matrix.org or #test:localhost:8448.
- .
- Matrix user IDs look like @matthew:matrix.org (although in the future
- you will normally refer to yourself and others using a 3PID: email
- address, phone number, etc rather than manipulating Matrix user IDs)
- .
- The overall architecture is:
- client <------> homeserver <=============> homeserver <------> client
-                 https://a.org/_matrix      https://b.net/_matrix
+ implementation.
```
Comment (richvdh) [inline debian/control]: trailing NL pls

File: debian/install [hunk 8]
```diff
@@ -1,2 +1,3 @@
-debian/homeserver.yaml etc/matrix-synapse
-debian/log.yaml etc/matrix-synapse
+../homeserver.yaml etc/matrix-synapse
```
Comment (richvdh) [inline debian/install]: are these not still inside the `debian` directory?

File: debian/install [hunk 9]
```diff
@@ -1,2 +1,3 @@
...
+../log.yaml etc/matrix-synapse
+../matrix-synapse.service lib/systemd/system
```
Comment (richvdh) [inline debian/install]: rather than do this, can we give the `--name=matrix-synapse` argument to `dh_installinit`, which will also mean that we can preserve the name of the `.default` file?

File: debian/matrix-synapse-py3.links [hunk 10]
```diff
@@ -0,0 +1,4 @@
+opt/venvs/matrix-synapse/bin/hash_password usr/bin/hash_password
+opt/venvs/matrix-synapse/bin/register_new_matrix_user usr/bin/register_new_matrix_user
+opt/venvs/matrix-synapse/bin/synapse_port_db usr/bin/synapse_port_db
+opt/venvs/matrix-synapse/bin/synctl usr/bin/synctl
```
Comment (richvdh) [inline debian/matrix-synapse-py3.links]: trailing NL please

File: debian/matrix-synapse-py3.triggers [hunk 11]
```diff
@@ -0,0 +1,9 @@
+# Register interest in Python interpreter changes (Python 2 for now); and
```
Comment (richvdh) [inline debian/matrix-synapse-py3.triggers]: > Python 2 for now

seems to be a bit of a lie.

Comment (richvdh) [inline debian/matrix-synapse-py3.triggers]: don't we have to do something in the postinst to handle the triggers?

File: debian/matrix-synapse-py3.triggers [hunk 12]
```diff
@@ -0,0 +1,9 @@
...
+# don't make the Python package dependent on the virtualenv package
+# processing (noawait)
+interest-noawait /usr/bin/python3.5
+interest-noawait /usr/bin/python3.6
+interest-noawait /usr/bin/python3.7
+
+# Also provide a symbolic trigger for all dh-virtualenv packages
```
Comment (richvdh) [inline debian/matrix-synapse-py3.triggers]: s/provide/listen to/?

File: debian/matrix-synapse.service [hunk 13]
```diff
@@ -6,10 +6,11 @@ Type=simple
...
-ExecStartPre=/usr/bin/python -m synapse.app.homeserver --config-path=/etc/matrix-synapse/homeserver.yaml --config-path=/etc/matrix-synapse/conf.d/ --generate-keys
-ExecStart=/usr/bin/python -m synapse.app.homeserver --config-path=/etc/matrix-synapse/homeserver.yaml --config-path=/etc/matrix-synapse/conf.d/
+ExecStartPre=/opt/venvs/matrix-synapse/bin/python -m synapse.app.homeserver --config-path=/etc/matrix-synapse/homeserver.yaml --config-path=/etc/matrix-synapse/conf.d/ --generate-keys
+ExecStart=/opt/venvs/matrix-synapse/bin/python -m synapse.app.homeserver --config-path=/etc/matrix-synapse/homeserver.yaml --config-path=/etc/matrix-synapse/conf.d/
 Restart=always
 RestartSec=3
 
 [Install]
 WantedBy=multi-user.target
+Alias=matrix-synapse
```
Comment (richvdh) [inline debian/matrix-synapse.service]: what's this doing?

File: debian/rules [hunk 14]
```diff
@@ -1,18 +1,51 @@
...
+	dh $@ $(DH_VENV_ARGS) --sourcedir $(SDIST_DIR)
 
-override_dh_auto_build:
+override_dh_gencontrol:
+	dh_gencontrol -- -v$(RELEASE_VERSION)
 
-override_dh_installinit:
-	dh_installinit --no-start
+override_dh_builddeb:
+	dh_builddeb -- -z9
```
Comment (richvdh) [inline debian/rules]: I'm not sure there's any value in this override. Let's let the default do its thing.

File: debian/rules [hunk 15]
```diff
@@ -1,18 +1,51 @@
...
+export DH_VIRTUALENV_INSTALL_ROOT=/opt/venvs
+export CFLAGS=-march=x86-64
+SNAKE=/usr/bin/python3
+DEB_BUILD_OPTIONS=noddebs
+EXTRA_REQUIREMENTS=--preinstall "setuptools>=17.1" --preinstall "pip>=7" --preinstall "wheel" --preinstall "lxml"
+DH_VENV_ARGS=--install-suffix "matrix-synapse" --with python-virtualenv --setuptools --python $(SNAKE) $(EXTRA_REQUIREMENTS) --extra-pip-arg="--no-cache-dir" --extra-pip-arg="--compile"
+PACKAGE=$(shell dh_listpackages)
+VERSION=$(shell $(SNAKE) setup.py --version)
+SDIST_DIR=debian/$(PACKAGE)-$(VERSION)
+RELEASE_VERSION=$(VERSION)~$(shell lsb_release -c -s)
```
Comment (richvdh) [inline debian/rules]: do we really need separate version numbers for different dists? it doesn't sound like a very debian thing to do to me.

File: debian/source/format [hunk 16]
```diff
@@ -1 +0,0 @@
-3.0 (quilt)
```
Comment (richvdh) [inline debian/source/format]: I'm sorry, I've still not grokked why this is going away. shouldn't it be `3.0 (native)` (https://www.debian.org/doc/manuals/maint-guide/dother.en.html#sourcef) or something?

File: debian/templates [hunk 17]
```diff
@@ -1,11 +1,11 @@
-Template: matrix-synapse/server-name
+Template: matrix-synapse-py3/server-name
```
Comment (richvdh) [inline debian/templates]: it feels like a mistake to rename the config settings. The existing config settings for the existing package can equally well apply to the new package, and it will save the user having to answer the question again when switching.

File: docker/Dockerfile-dhvirtualenv [hunk 18]
```diff
@@ -0,0 +1,9 @@
+ARG distro="debian:stable"
```
Comment (richvdh) [inline docker/Dockerfile-dhvirtualenv]: A comment about what this does would be helpful.

File: docker/Dockerfile-dhvirtualenv [hunk 19]
```diff
@@ -0,0 +1,9 @@
...
+FROM ${distro} AS dpkg-build
+RUN apt-get update -qq -o Acquire::Languages=none \
+    && env DEBIAN_FRONTEND=noninteractive apt-get install \
+        -yqq --no-install-recommends -o Dpkg::Options::=--force-unsafe-io \
+        build-essential debhelper devscripts equivs dh-virtualenv libpq-dev sqlite3 python3-dev libffi-dev libjpeg-dev zlib1g-dev libxslt1-dev python3-setuptools python3-pip libssl-dev lsb-release
+WORKDIR /synapse/build
+ENTRYPOINT ["bash","/synapse/build/debian/build_debian.sh"]
+
```
Comment (richvdh) [inline docker/Dockerfile-dhvirtualenv]: this one actually has a spurious trailing NL fwiw ;).

File: docker/build_debian_packages.sh [hunk 20]
```diff
@@ -0,0 +1,11 @@
...
+
+mkdir -p ../debs
+
+docker build --tag dh-venv-builder:bionic --build-arg distro=ubuntu:bionic -f Dockerfile-dhvirtualenv .
+#docker build --tag dh-venv-builder:stretch --build-arg distro=debian:stretch -f Dockerfile-dhvirtualenv .
+
+docker run -it --rm --volume=$(pwd)/../\:/synapse/build dh-venv-builder:bionic
+#ocker run -it --rm --volume=$(pwd)/../\:/build dh-venv-builder:stretch
+
+sudo chown $(id -u):$(id -g) -R ../debs/
```
Comment (richvdh) [inline docker/build_debian_packages.sh]: trailing NL please

File: debian/build_debian.sh (hunk 2 above)
Comment (richvdh) [reply debian/build_debian.sh]: (and I'd prefer scripts that exist solely to be run in docker images to live in the `docker` directory)

File: docker/build_debian_packages.sh [hunk 21]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
```
Comment (richvdh) [inline docker/build_debian_packages.sh]: can has comment pls?

File: docker/build_debian_packages.sh [hunk 22]
```diff
@@ -0,0 +1,11 @@
...
+
+mkdir -p ../debs
+
+docker build --tag dh-venv-builder:bionic --build-arg distro=ubuntu:bionic -f Dockerfile-dhvirtualenv .
+#docker build --tag dh-venv-builder:stretch --build-arg distro=debian:stretch -f Dockerfile-dhvirtualenv .
+
+docker run -it --rm --volume=$(pwd)/../\:/synapse/build dh-venv-builder:bionic
+#ocker run -it --rm --volume=$(pwd)/../\:/build dh-venv-builder:stretch
```
Comment (richvdh) [inline docker/build_debian_packages.sh]: I think if we're going to keep the alternatives around, we should maintain them:

```suggestion
#docker run -it --rm --volume=$(pwd)/../\:/synapse/build dh-venv-builder:stretch
```

is it right that it

File: debian/install (hunk 9 above)
Comment (richvdh) [reply debian/install]: (i'm also not sure that /lib is the right place for this file, but that becomes moot if we do it with installinit instead)

File: .dockerignore [hunk 23]
```diff
@@ -5,3 +5,5 @@ demo/etc
 tox.ini
 .git/*
 .tox/*
+debian/matrix-synapse/
+debian/matrix-synapse-*/
```
Comment (richvdh) [inline .dockerignore]: trailing NL please

File: debian/rules (hunk 15 above)
Comment (richvdh) [reply debian/rules]: (I'm also wondering if `~` is the right separator here, though I guess it is as good as any other)

File: debian/rules [hunk 24]
```diff
@@ -1,18 +1,51 @@
...
-	python setup.py install --root=debian/matrix-synapse --install-layout=deb
+build-arch:
+	$(SNAKE) setup.py sdist --formats tar
+	mkdir -p $(SDIST_DIR)
+	tar -x -C $(SDIST_DIR) --strip-components=1 --exclude '*.egg-info' -f dist/*.tar
+	dh $@ $(DH_VENV_ARGS) --sourcedir $(SDIST_DIR)
 
-override_dh_auto_build:
+override_dh_gencontrol:
+	dh_gencontrol -- -v$(RELEASE_VERSION)
```
Comment (richvdh) [inline debian/rules]: can we do this by messing around with `dch` (presumably before firing up `dpkg_buildpackage`) instead? My experience is that the debian tooling will get confused if the package version doesn't match the latest in `debian/changelog`.

Comment (hawkowl) [reply debian/rules]: I haven't noticed any problems thus far, and it appears to be a supported method.

File: debian/control (hunk 6 above)
Comment (richvdh) [reply debian/control]: actually not this, because afaict it was only used by the init script, which has gone away

File: debian/install (hunk 8 above)
Comment (hawkowl) [reply debian/install]: so, the old system would copy the entire tree into the source dir, and so it would be. This one only extracts a tarball, so instead of getting it from the 'source' we need to go back down into the original debian directory and fish it out.

File: debian/install (hunk 9 above)
Comment (hawkowl) [reply debian/install]: `/lib` is the right spot, it's symlinked into `/etc` by `systemctl enable`.

File: debian/matrix-synapse-py3.triggers (hunk 11 above)
Comment (hawkowl) [reply debian/matrix-synapse-py3.triggers]: that is true :thinking:

File: debian/rules (hunk 15 above)
Comment (hawkowl) [reply debian/rules]: It's what I've seen before, and we do need to make different debs for different diffs (because targeting different Python 3 versions). https://www.debian.org/doc/debian-policy/ch-controlfields.html#s-f-version says it looks fine, I think

File: debian/matrix-synapse-py3.triggers (hunk 12 above)
Comment (richvdh) [reply debian/matrix-synapse-py3.triggers]: well, let's not just cargo-cult stuff without understanding it please...

File: debian/source/format (hunk 16 above)
Comment (richvdh) [reply debian/source/format]: I'm not sure that's relevant. `debian/source/format` is a standard thing for any debian packaging aiui

File: debian/install (hunk 9 above)
Comment (richvdh) [reply debian/install]: ok the docs on https://manpages.debian.org/stretch/debhelper/dh_installinit.1.en.html convince me that /lib is the right spot, but not that we should be doing it this way.

File: debian/rules (hunk 15 above)
Comment (richvdh) [reply debian/rules]: ok, seems fair.

File: debian/rules [hunk 25]
```diff
@@ -1,18 +1,50 @@
...
-export PYBUILD_NAME=matrix-synapse
-%:
-	dh $@ --with python2 --with systemd --buildsystem=pybuild --no-guessing-deps
+# Increase trace logging, see debhelper(7) (uncomment to enable)
+#DH_VERBOSE=1
+
+export DH_VIRTUALENV_INSTALL_ROOT=/opt/venvs
+export CFLAGS=-march=x86-64
+SNAKE=/usr/bin/python3
+DEB_BUILD_OPTIONS=noddebs
```
Comment (richvdh) [inline debian/rules]: what does this do?

File: debian/rules [hunk 26]
```diff
@@ -1,18 +1,50 @@
...
-# This file was automatically generated by stdeb 0.8.2 at
-# Fri, 12 Jun 2015 14:32:03 +0100
-export PYBUILD_NAME=matrix-synapse
-%:
-	dh $@ --with python2 --with systemd --buildsystem=pybuild --no-guessing-deps
+# Increase trace logging, see debhelper(7) (uncomment to enable)
+#DH_VERBOSE=1
+
+export DH_VIRTUALENV_INSTALL_ROOT=/opt/venvs
+export CFLAGS=-march=x86-64
```
Comment (richvdh) [inline debian/rules]: why is this required?

File: debian/rules (hunk 25 above)
Comment (richvdh) [reply debian/rules]: (I think it's a deprecated option for dh_strip, but since we're disabling dh_strip anyway, it seems redundant?)

File: debian/rules [hunk 27]
```diff
@@ -1,18 +1,50 @@
...
+EXTRA_REQUIREMENTS=--preinstall "setuptools>=17.1" --preinstall "pip>=7" --preinstall "wheel" --preinstall "lxml"
+DH_VENV_ARGS=--install-suffix "matrix-synapse" --with python-virtualenv  --builtin-venv --setuptools --python $(SNAKE) $(EXTRA_REQUIREMENTS) --extra-pip-arg="--no-cache-dir" --extra-pip-arg="--compile"
+PACKAGE=$(shell dh_listpackages)
+VERSION=$(shell $(SNAKE) setup.py --version)
+SDIST_DIR=debian/$(PACKAGE)-$(VERSION)
+RELEASE_VERSION=$(VERSION)
+DEBIAN_RELEASE_NUMBER=1
+DISTRO_RELEASE=$(shell lsb_release -c -s)
+
+clean:
```
Comment (richvdh) [inline debian/rules]: can we not provide `override_dh_clean` rather than overriding the whole of `clean` ?

File: debian/control [hunk 28]
```diff
@@ -1,77 +1,41 @@
...
+ libffi-dev,
+ libjpeg-dev,
+ libssl-dev,
+ libxslt1-dev,
+ lsb-release,
+ python3-dev,
+ python3,
+ python3-setuptools,
+ python3-pip,
+ python3-venv,
```
Comment (richvdh) [inline debian/control]: where do we actually use this (as opposed to `python-virtualenv`)?

File: debian/control [hunk 29]
```diff
@@ -1,77 +1,41 @@
...
- python-lxml,
+ libffi6,
+ libjpeg-turbo8,
+ libpq5,
+ libxslt1.1,
+ openssl,
+ python3-distutils,
+ python3,
+ sqlite3,
+ ${misc:Pre-Depends}
```
Comment (richvdh) [inline debian/control]: should this not be ${misc:Depends}, given it's in the Depends section?

File: debian/control [hunk 30]
```diff
@@ -1,77 +1,41 @@
...
 
-Package: matrix-synapse
-Architecture: all
+Package: matrix-synapse-py3
+Architecture: amd64
+Conflicts: matrix-synapse
+Pre-Depends: dpkg (>= 1.16.1)
 Depends:
- ${misc:Depends},
- ${python:Depends},
```
Comment (richvdh) [inline debian/control]: do we not need this? dh_virtualenv's man page suggests we do

File: debian/rules [hunk 31]
```diff
@@ -1,18 +1,50 @@
 #!/usr/bin/make -f
+#
+# Build Debian package using https://github.com/spotify/dh-virtualenv
+#
+# The below targets create a clean copy of the workdir via
```
Comment (richvdh) [inline debian/rules]: I'm afraid I'm completely failing to grok this. why does the presence of the `debian` dir make it go haywire?

(`pip install .` works just fine for me)

File: debian/rules [hunk 32]
```diff
@@ -1,18 +1,50 @@
...
-%:
-	dh $@ --with python2 --with systemd --buildsystem=pybuild --no-guessing-deps
+# Increase trace logging, see debhelper(7) (uncomment to enable)
+#DH_VERBOSE=1
+
+export DH_VIRTUALENV_INSTALL_ROOT=/opt/venvs
+export CFLAGS=-march=x86-64
+SNAKE=/usr/bin/python3
+DEB_BUILD_OPTIONS=noddebs
+EXTRA_REQUIREMENTS=--preinstall "setuptools>=17.1" --preinstall "pip>=7" --preinstall "wheel" --preinstall "lxml"
```
Comment (richvdh) [inline debian/rules]: can you break this into multiple lines with `\`?

File: debian/rules [hunk 33]
```diff
@@ -1,18 +1,50 @@
...
+DH_VENV_ARGS=--install-suffix "matrix-synapse" --with python-virtualenv  --builtin-venv --setuptools --python $(SNAKE) $(EXTRA_REQUIREMENTS) --extra-pip-arg="--no-cache-dir" --extra-pip-arg="--compile"
```
Comment (richvdh) [inline debian/rules]: can you break this up too?

Comment (richvdh) [inline debian/rules]: given that most of these only apply to `dh_virtualenv`, can we put them in an `override_dh_virtualenv` target rather than passing them to every single debhelper script?

File: debian/rules [hunk 34]
```diff
@@ -1,18 +1,50 @@
...
+DISTRO_RELEASE=$(shell lsb_release -c -s)
+
+clean:
+	test ! -d dist || rm -rf dist
+	test ! -d $(SDIST_DIR) || rm -rf $(SDIST_DIR)
+	dh $@ $(DH_VENV_ARGS)
 
-override_dh_auto_install:
-	python setup.py install --root=debian/matrix-synapse --install-layout=deb
+build-arch:
```
Comment (richvdh) [inline debian/rules]: overriding the whole of `build-arch` seems pretty brutal. can we just put it in an `override_dh_virtualenv` ?

File: debian/rules (hunk 24 above)
Comment (richvdh) [reply debian/rules]: ok fair.

File: debian/rules (hunk 26 above)
Comment (hawkowl) [reply debian/rules]: So that things like Twisted with C extensions compile with a generic compiler flag set -- the generic x86-64 one, not like `haswell`.

File: debian/control (hunk 28 above)
Comment (hawkowl) [reply debian/control]: `python-virtualenv` is broken on sid, so we use `--builtin-venv` so dh_virtualenv uses the built in one. Debian breaks it up into its own package, because Debian.

File: debian/control (hunk 29 above)
Comment (hawkowl) [reply debian/control]: yep

File: debian/control (hunk 30 above)
Comment (hawkowl) [reply debian/control]: their packaging cookbook doesn't use it...

File: debian/rules (hunk 31 above)
Comment (hawkowl) [reply debian/rules]: I'll remove it, this was from the cookiecutter, it looks like modern pip is more sensible.

File: debian/rules (hunk 34 above)
Comment (hawkowl) [reply debian/rules]: I guess we don't need it all, if the tarballing works.

File: debian/rules (hunk 33 above)
Comment (hawkowl) [reply debian/rules]: sure

File: debian/install (hunk 8 above)
Comment (hawkowl) [reply debian/install]: I think I fixed it!

File: debian/matrix-synapse-py3.triggers (hunk 12 above)
Comment (hawkowl) [reply debian/matrix-synapse-py3.triggers]: well i'm also cargo culting it from the dh_virtualenv docs, I have no idea what it means in a real sense.

File: debian/rules (hunk 26 above)
Comment (richvdh) [reply debian/rules]: I see. could you add a comment to say so please?

File: debian/rules (hunk 34 above)
Comment (richvdh) [reply debian/rules]: even better \o/

File: debian/rules (hunk 33 above)
Comment (richvdh) [reply debian/rules]: yay

File: debian/rules (hunk 31 above)
Comment (richvdh) [reply debian/rules]: \o/

File: debian/matrix-synapse-py3.triggers (hunk 12 above)
Comment (richvdh) [reply debian/matrix-synapse-py3.triggers]: Sigh; I see what you mean. Having stared at the docs, the dh-virtualenv source, the generated postinst... I also have no idea what they are trying to achieve with this.

File: debian/rules [hunk 35]
```diff
@@ -1,18 +1,38 @@
...
-# This file was automatically generated by stdeb 0.8.2 at
-# Fri, 12 Jun 2015 14:32:03 +0100
-export PYBUILD_NAME=matrix-synapse
-%:
-	dh $@ --with python2 --with systemd --buildsystem=pybuild --no-guessing-deps
+export DH_VIRTUALENV_INSTALL_ROOT=/opt/venvs
+export CFLAGS=-march=x86-64
+SNAKE=/usr/bin/python3
+DH_VENV_ARGS=--install-suffix "matrix-synapse" \
+			 --builtin-venv\
```
Comment (richvdh) [inline debian/rules]: could you add a comment somewhere here that explains a little bit about why we use builtin-venv?

File: debian/control [hunk 36]
```diff
@@ -34,7 +34,8 @@ Depends:
  python3-distutils,
  python3,
  sqlite3,
- ${misc:Pre-Depends}
+ ${perl:Depends},
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply debian/control]: dh complains I haven't used it, so I added it. :S

File: debian/rules [hunk 37]
```diff
@@ -2,41 +2,26 @@
...
+			 --setuptools \
+			 --python $(SNAKE) \
+			 --preinstall "setuptools>=40.6.3" \
+			 --preinstall "pip>=18" \
+			 --preinstall "wheel" \
+			 --extra-pip-arg="--no-cache-dir" \
+			 --extra-pip-arg="--compile"
 PACKAGE=$(shell dh_listpackages)
 VERSION=$(shell $(SNAKE) setup.py --version)
 SDIST_DIR=debian/$(PACKAGE)-$(VERSION)
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply debian/rules]: yep!

File: synapse/python_dependencies.py [hunk 38]
```diff
@@ -84,6 +84,9 @@
     "saml2": {
         "pysaml2>=4.5.0": ["saml2"],
     },
+    "url_preview": {
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply synapse/python_dependencies.py]: OK, reverted.

File: debian/rules (hunk 35 above)
Comment (hawkowl) [reply debian/rules]: done

File: debian/matrix-synapse-py3.triggers (hunk 12 above)
Comment (hawkowl) [reply debian/matrix-synapse-py3.triggers]: debiaaaaaaaaaaaaaaaaaaaaaaaaaaaaaan :D

File: debian/install (hunk 9 above)
Comment (hawkowl) [reply debian/install]: fwiw we can do it the "right way" when we rename the package, I didn't find a way for dh_systemd to install it under a name that wasn't the package name. :/ Once it's just `matrix-synapse` again, we'll be able to use dh_systemd and it'll suck less.

File: debian/rules (hunk 25 above)
Comment (hawkowl) [reply debian/rules]: removed

File: debian/rules (hunk 26 above)
Comment (hawkowl) [reply debian/rules]: done

File: debian/install [hunk 39]
```diff
@@ -1,2 +1,3 @@
 debian/homeserver.yaml etc/matrix-synapse
 debian/log.yaml etc/matrix-synapse
+debian/matrix-synapse.service lib/systemd/system
```
Comment (richvdh) [inline debian/install]: > fwiw we can do it the "right way" when we rename the package, I didn't find a way for dh_systemd to install it under a name that wasn't the package name. :/

I'm totally failing to follow your logic here. We aren't using dh_systemd* at all currently (because we are on debhelper compat 9, which doesn't enable it by default, and instead dh_installinit`does the installation of the service file).

Either way, the solution is just to add a `--name` argument in the relevant `override_dh_*` sections.

File: debian/matrix-synapse.service [hunk 40]
```diff
@@ -6,8 +6,8 @@ Type=simple
 User=matrix-synapse
 WorkingDirectory=/var/lib/matrix-synapse
 EnvironmentFile=/etc/default/matrix-synapse
```
Comment (richvdh) [inline debian/matrix-synapse.service]: (putting this comment here for want of anywhere else to put it):

we're currently installing the default file as `/etc/default/matrix-synapse-py3`, so this won't work. We shouldn't rename it, because that will drop any local changes that people have made.

File: debian/control [hunk 41]
```diff
@@ -5,10 +5,6 @@ Maintainer: Synapse Developers <synapse@matrix.org>
 Build-Depends:
  debhelper (>= 9),
  dh-virtualenv (>= 0.10),
- libffi-dev,
```
[the remark this answers is not in the document]
Comment (richvdh) [reply debian/control]: (you explained this already: because we're using static libs)

File: docker/Dockerfile-dhvirtualenv [hunk 42]
```diff
@@ -10,26 +10,20 @@ RUN apt-get update -qq -o Acquire::Languages=none \
...
         python3-dev \
         python3-pip \
         python3-setuptools \
         python3-venv \
         sqlite3 \
-        wget \
-        zlib1g-dev
+        wget
 
 RUN mkdir -p /tmp/dhvenv \
```
[the remark this answers is not in the document]
Comment (richvdh) [reply docker/Dockerfile-dhvirtualenv]: (you've moved this out of the dockerfile, so nm)

File: debian/rules [hunk 43]
```diff
@@ -32,6 +36,7 @@ override_dh_shlibdeps:
 
 override_dh_virtualenv:
 	dh_virtualenv $(DH_VENV_ARGS)
+	debian/matrix-synapse-py3/opt/venvs/matrix-synapse/bin/python -B -m twisted.trial --reporter=text -j 2 ../tests/
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply debian/rules]: because our current working dir is the `debian/` directory

[the remark this answers is not in the document]
Comment (richvdh) [reply debian/rules]: I don't think it is?

[the remark this answers is not in the document]
Comment (richvdh) [reply debian/rules]: (if it is, why is the command `*debian*/matrix-synapse-py3/opt/venvs/matrix-synapse/bin/python`?)

File: debian/source/format (hunk 16 above)
Comment (hawkowl) [reply debian/source/format]: not for ones without source packages, as far as I can tell?

File: debian/rules (hunk 43 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply debian/rules]: ok, the reason this works is because [twisted.python.reflect.filenameToModuleName](https://github.com/twisted/twisted/blob/trunk/src/twisted/python/reflect.py#L325) falls back to returning the final component of the path (`tests` in this case) if it can't find a matching python package anywhere else.

File: debian/NEWS [hunk 44]
```diff
@@ -0,0 +1,22 @@
...
+    instead of using the system python libraries. (This may mean that you can
+    remove a number of old dependencies with `apt-get autoremove`).
+
+  matrix-synapse-py3 will take over responsibility for the existing
+  configuration files, including the matrix-synapse systemd service.
+
+  Beware, however, that `apt-get purge matrix-synapse` will *disable* the
+  matrix-synapse service (so that it will not be started on reboot), even
+  though that service is no longer being provided by the matrix-synapse
+  package. It can be re-enabled with `systemctl matrix-synapse`.
```
Comment (uhoreg) [inline debian/NEWS]: Should be "`systemctl enable matrix-synapse`"?

File: debian/rules [hunk 45]
```diff
@@ -1,18 +1,49 @@
...
-%:
-	dh $@ --with python2 --with systemd --buildsystem=pybuild --no-guessing-deps
-
-override_dh_auto_install:
-	python setup.py install --root=debian/matrix-synapse --install-layout=deb
+export DH_VIRTUALENV_INSTALL_ROOT=/opt/venvs
+# Use -march=x86-64 so any compiled C extensions are compiled with the most
+# generic as possible x64 instructions, so that compiling it on a new Intel chip
+# doesn't enable features not available on older ones or AMD.
+export CFLAGS=-march=x86-64
```
Comment (uhoreg) [inline debian/rules]: Surely this will break on non-x86-64 architectures (like ARM or i386).  Unless we're only planning on ever supporting x86-64, in which case, \<shrug\>

File: debian/NEWS (hunk 44 above)
Comment (richvdh) [reply debian/NEWS]: indeed.

File: debian/rules (hunk 45 above)
Comment (richvdh) [reply debian/rules]: currently, we're only ever planning on shipping packages for x86-64, though I'm inclined to agree it's a shame that this will stop anyone else using our build scripts to build packages for other archs.

I'm not sure that I can think of a better way though?

Comment (hawkowl) [reply debian/rules]: Other than having that set in the environment using `-e CFLAGS="-march=x86_64"` in the `docker run`, and testing that it gets passed through. That would probably work fine.

Comment (uhoreg) [reply debian/rules]: One option would be to check the output of `dpkg-architecture -q DEB_HOST_ARCH`, and only set that if it's equal to `amd64`

Comment (richvdh) [reply debian/rules]: ok I've moved a bunch of this stuff out to a separate shell script, and made the CFLAGS conditional on `dpkg-architecture -q DEB_HOST_ARCH` as per @uhoreg 's suggestion. thanks!
